File: app/src/main/java/com/rsf/innometrics/ViewAdapter.java
package com.rsf.innometrics;

import android.annotation.SuppressLint;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

public class ViewAdapter extends RecyclerView.Adapter<ViewAdapter.ViewHolder> {

    private List<MappingApp> mMappingAppList = new ArrayList<>();
    @SuppressLint("SimpleDateFormat")
    private DateFormat mDateFormat = new SimpleDateFormat();

    static class ViewHolder extends RecyclerView.ViewHolder {
        private final TextView mPackageName;
        private final TextView mLastTimeUsed;
        private final ImageView mAppIcon;

        ViewHolder(View v) {
            super(v);
            mPackageName = v.findViewById(R.id.textview_package_name);
            mLastTimeUsed = v.findViewById(R.id.textview_last_time_used);
            mAppIcon = v.findViewById(R.id.app_icon);
        }

        TextView getLastTimeUsed() {
            return mLastTimeUsed;
        }

        TextView getPackageName() {
            return mPackageName;
        }

        ImageView getAppIcon() {
            return mAppIcon;
        }
    }

    ViewAdapter() {
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(ViewGroup viewGroup, int viewType) {
        View v = LayoutInflater.from(viewGroup.getContext())
                .inflate(R.layout.usage_row, viewGroup, false);
        return new ViewHolder(v);
    }

    @Override
    public void onBindViewHolder(ViewHolder viewHolder, final int position) {
        viewHolder.getPackageName().setText(
                mMappingAppList.get(position).usageStats.getPackageName());
        long lastTimeUsed = mMappingAppList.get(position).usageStats.getLastTimeUsed();
        viewHolder.getLastTimeUsed().setText(mDateFormat.format(new Date(lastTimeUsed)));
        viewHolder.getAppIcon().setImageDrawable(mMappingAppList.get(position).appIcon);
    }

    @Override
    public int getItemCount() {
        return mMappingAppList.size();
    }

    void setCustomUsageStatsList(List<MappingApp> customUsageStats) {
        mMappingAppList = customUsageStats;
    }
}

File: app/src/main/java/com/rsf/innometrics/ViewAdapter.kt
package com.rsf.innometrics

import android.content.Context
import android.content.pm.PackageManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.LiveData
import androidx.recyclerview.widget.RecyclerView
import com.rsf.innometrics.vo.Stats
import java.text.DateFormat
import java.util.*

class ViewAdapter internal constructor() : RecyclerView.Adapter<ViewAdapter.ViewHolder>() {
    private var mAppStatsList: List<Stats> = emptyList()
    private lateinit var activity:FragmentActivity
    private val mDateFormat = DateFormat.getDateTimeInstance()

    class ViewHolder(v: View) : RecyclerView.ViewHolder(v) {
        val appName: TextView = v.findViewById(R.id.textview_app_name)
        val lastTimeUsed: TextView = v.findViewById(R.id.textview_total_time_used)
        val appIcon: ImageView = v.findViewById(R.id.app_icon)
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, viewType: Int): ViewHolder {
        val v = LayoutInflater.from(viewGroup.context)
                .inflate(R.layout.usage_row, viewGroup, false)
        return ViewHolder(v)
    }

    override fun onBindViewHolder(viewHolder: ViewHolder, position: Int) {
        val (_, name, time_begin, time_end) = mAppStatsList[position]
        val time = mDateFormat.format(Date(time_end))
        val label =
            activity.packageManager.getApplicationLabel(
                    activity.packageManager.getApplicationInfo(
                            name, PackageManager.GET_META_DATA))
        val icon = try {
            activity.packageManager.getApplicationIcon(name)
        } catch (e: PackageManager.NameNotFoundException) {
            activity.getDrawable(
                    R.drawable.ic_android_black_24dp)
        }
        viewHolder.appName.text = label
        viewHolder.appIcon.setImageDrawable(icon)
        viewHolder.lastTimeUsed.text = time
    }

    override fun getItemCount(): Int {
        return mAppStatsList.size
    }

    fun updateStatsList(customUsageStats: List<Stats>, activity: FragmentActivity) {
        mAppStatsList = customUsageStats
        this.activity = activity
        notifyDataSetChanged()
    }
}